File: ios/IsekaiCameraClient/App/ViewerSettings.swift
import Foundation

/// Everything the connect screen collects apart from the secrets, which live in
/// the keychain. Persisted so a dev round-trip does not mean retyping five
/// fields; the defaults match the desktop `camera-client`.
struct ViewerSettings: Codable, Equatable {
    var identityURL = "https://identity.isekai.tools:9443"
    var proxyURL = "https://tokyo.link.isekai.tools:8443"
    var protocolName = "isekai-validator-v1"
    /// Issued by the camera server for this viewer's Endpoint ID.
    var capability = ""
    /// The camera server's "Listener ID".
    var listenerID = ""
    /// The Endpoint that camera is expected to be, kept beside the listener it
    /// is running. Saved rather than re-read because the check it feeds is
    /// against the proxy, and a relaunch must not quietly drop it. Empty for a
    /// camera reached by a hand-carried capability.
    var expectedEndpoint = ""
    /// Register the Endpoint with the Identity API before a token is issued.
    var register = true
    /// **Dev only.** Accepts self-signed proxy/Identity certificates; Phase 4 of
    /// the plan turns real validation back on.
    var insecureSkipVerify = false
    /// Offer a direct path and allow migrating off the relay. Off makes the
    /// session relay-only, which is also how to tell a migration problem apart
    /// from a relay one.
    var enableMigration = true
    /// `RUST_LOG`-style filter for the core's logging. Empty disables it.
    /// There is no console on a phone, so this is the only way to see what the
    /// core is doing.
    var logFilter = ""
}

extension ViewerSettings {
    private enum CodingKeys: String, CodingKey {
        case identityURL, proxyURL, protocolName, capability, listenerID
        case expectedEndpoint, register, insecureSkipVerify, enableMigration, logFilter
    }

    /// Decoded a field at a time, so a blob written by an older build keeps
    /// what it does have.
    ///
    /// A property's default value does not cover it: the synthesized decoder
    /// throws `keyNotFound` for a field added since, and `load()` falls back to
    /// defaults **wholesale** — so adding one field would have reset every
    /// server URL and dropped the selected camera on every installed app.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let d = ViewerSettings()
        identityURL = try c.decodeIfPresent(String.self, forKey: .identityURL) ?? d.identityURL
        proxyURL = try c.decodeIfPresent(String.self, forKey: .proxyURL) ?? d.proxyURL
        protocolName = try c.decodeIfPresent(String.self, forKey: .protocolName) ?? d.protocolName
        capability = try c.decodeIfPresent(String.self, forKey: .capability) ?? d.capability
        listenerID = try c.decodeIfPresent(String.self, forKey: .listenerID) ?? d.listenerID
        expectedEndpoint =
            try c.decodeIfPresent(String.self, forKey: .expectedEndpoint) ?? d.expectedEndpoint
        register = try c.decodeIfPresent(Bool.self, forKey: .register) ?? d.register
        insecureSkipVerify =
            try c.decodeIfPresent(Bool.self, forKey: .insecureSkipVerify) ?? d.insecureSkipVerify
        enableMigration =
            try c.decodeIfPresent(Bool.self, forKey: .enableMigration) ?? d.enableMigration
        logFilter = try c.decodeIfPresent(String.self, forKey: .logFilter) ?? d.logFilter
    }

    private static let defaultsKey = "viewer.settings"

    /// The saved settings, or fresh defaults.
    static func load() -> ViewerSettings {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey),
              let decoded = try? JSONDecoder().decode(ViewerSettings.self, from: data)
        else { return ViewerSettings() }
        return decoded
    }

    func save() {
        guard let data = try? JSONEncoder().encode(self) else { return }
        UserDefaults.standard.set(data, forKey: Self.defaultsKey)
    }
}

File: android/app/src/main/kotlin/tools/isekai/cameraclient/ViewerSettings.kt
package tools.isekai.cameraclient

import android.content.Context
import android.content.SharedPreferences

/**
 * Everything the connect screen collects apart from the secrets, which live
 * in [SecureStore]. Persisted so a dev round-trip does not mean retyping
 * fields; the defaults match the desktop `camera-client` and the iOS app's
 * own `ViewerSettings`.
 */
data class ViewerSettings(
    var identityUrl: String = "https://identity.isekai.tools:9443",
    var proxyUrl: String = "https://link.isekai.tools:6443",
    var protocolName: String = "isekai-validator-v1",
    var capability: String = "",
    var listenerId: String = "",
    // Which Endpoint `listenerId` is expected to be, from the camera it was
    // picked from -- kept so connecting can hold the proxy's answer against
    // it (`camera_core::paired`). Empty for a hand-typed listener, which
    // brings no pairing to check against. Mirrors the iOS app's own
    // `expectedEndpoint`.
    var expectedEndpoint: String = "",
    var register: Boolean = true,
    // Dev only. Accepts self-signed proxy/Identity certificates. Defaults to
    // false now that IsekaiApplication installs a real CA bundle
    // (SSL_CERT_FILE) -- see ~/isekai-link-android-camera-client-setup.md,
    // task 7, for the bug this used to work around.
    var insecureSkipVerify: Boolean = false,
    var enableMigration: Boolean = true,
    var logFilter: String = "",
) {
    companion object {
        private const val PREFS_NAME = "viewer_settings"

        fun load(context: Context): ViewerSettings {
            val prefs = prefs(context)
            val defaults = ViewerSettings()
            return ViewerSettings(
                identityUrl = prefs.getString("identityUrl", defaults.identityUrl)!!,
                proxyUrl = prefs.getString("proxyUrl", defaults.proxyUrl)!!,
                protocolName = prefs.getString("protocolName", defaults.protocolName)!!,
                capability = prefs.getString("capability", defaults.capability)!!,
                listenerId = prefs.getString("listenerId", defaults.listenerId)!!,
                expectedEndpoint =
                    prefs.getString("expectedEndpoint", defaults.expectedEndpoint)!!,
                register = prefs.getBoolean("register", defaults.register),
                insecureSkipVerify =
                    prefs.getBoolean("insecureSkipVerify", defaults.insecureSkipVerify),
                enableMigration = prefs.getBoolean("enableMigration", defaults.enableMigration),
                logFilter = prefs.getString("logFilter", defaults.logFilter)!!,
            )
        }

        private fun prefs(context: Context): SharedPreferences =
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    fun save(context: Context) {
        prefs(context).edit()
            .putString("identityUrl", identityUrl)
            .putString("proxyUrl", proxyUrl)
            .putString("protocolName", protocolName)
            .putString("capability", capability)
            .putString("listenerId", listenerId)
            .putString("expectedEndpoint", expectedEndpoint)
            .putBoolean("register", register)
            .putBoolean("insecureSkipVerify", insecureSkipVerify)
            .putBoolean("enableMigration", enableMigration)
            .putString("logFilter", logFilter)
            .apply()
    }
}
